contracts/0.4.24/oracle/Algorithm.sol

```diff
@@ -11,55 +11,81 @@ library Algorithm {
     using SafeMath for uint256;
 
     /**
-      * Computes mode of a non-empty array, if array is unimodal.
-      * Low gas cost.
+      * @dev Finds the first element in the given array |data| that occures
+      * more that |quorum| times. Low gas cost for majority case. Returns the
+      * discovered element or 0 if no element is frequent enough.
       */
-    function mode(uint256[] data) internal pure returns (bool, uint256) {
-
+    function frequent(uint256[] data, uint256 quorum) internal pure returns (bool exists, uint256 mode) {
         assert(0 != data.length);
 
-        // allocate arrays
-        uint256[] memory dataValues = new uint256[](data.length);
-        uint256[] memory dataValuesCounts = new uint256[](data.length);
+        uint256 acc = 0;
+        uint256 ctr = 0;
+        uint256 i;
+        if (quorum * 2 > data.length) {
+            
+            // If the expected quorum is more then half of data length, apply
+            // Boyer–Moore majority vote algorithm
+            for (i = 0; i < data.length; ++i) {
+                if (ctr == 0) {
+                    acc = data[i];
+                    ++ctr;
+                } else if (acc == data[i]) {
+                    // If enough same elemnts in a row, immediately return them
+                    if (++ctr == quorum) return (true, acc);
+                } else {
+                    --ctr;
+                }
+            }
 
-        // initialize first element
-        dataValues[0] = data[0];
-        dataValuesCounts[0] = 1;
-        uint256 dataValuesLength = 1;
+            // And make sure the resulted element is frequent enough
+            ctr = 0;
+            for (i = 0; i < data.length; ++i) {
+                if (data[i] == acc && ++ctr == quorum) return (true, acc);
+            }
+        } else {
+            
+            // Otherwise, apply optimized insertion sort
```

Comment:
Btw, there was a quick sort impl here https://github.com/lidofinance/lido-dao/blob/8b1203b248b638275f441fc16de0f0e2e1893c55/apps/depooloracle/contracts/Algorithm.sol#L25-L65

Reply:
Cool! Maybe It would be good to put our algorithms in a separate library (like this one)?..
In this case, I chose insertion sort because `data.length` should be low. But as we discussed with @vshvsh, this branch (line 45 and onwards) should work extremely rarely so the gas price is not important here at all.

Reply:
I agree. It's completely up to you.

Reply:
Ok, will have in mind for now, thank you for pointing this out!